File: Object Pascal/VCL/Sensors/uMain.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit uMain;

interface

uses
  Winapi.Windows, Winapi.Messages, System.SysUtils, System.Variants, System.Classes, Vcl.Graphics,
  Vcl.Controls, Vcl.Forms, Vcl.Dialogs, Vcl.StdCtrls, System.Sensors,
  System.Sensors.Components, Vcl.ExtCtrls;

type
  TForm5 = class(TForm)
    GroupBox1: TGroupBox;
    GroupBox2: TGroupBox;
    GroupBox3: TGroupBox;
    lsLocation: TLocationSensor;
    msAccelerometr: TMotionSensor;
    osCompass: TOrientationSensor;
    osInclinometer: TOrientationSensor;
    tOrientation: TTimer;
    lCompass: TLabel;
    lInclinometr: TLabel;
    Button1: TButton;
    GroupBox4: TGroupBox;
    Panel1: TPanel;
    bSwitch: TButton;
    lCoordinates: TLabel;
    lAddress: TLabel;
    lGeoCoordinates: TLabel;
    tMotion: TTimer;
    lAccel: TLabel;
    procedure osCompassSensorChoosing(Sender: TObject;
      const Sensors: TSensorArray; var ChoseSensorIndex: Integer);
    procedure osInclinometerSensorChoosing(Sender: TObject;
      const Sensors: TSensorArray; var ChoseSensorIndex: Integer);
    procedure tOrientationTimer(Sender: TObject);
    procedure bSwitchClick(Sender: TObject);
    procedure lsLocationLocationChanged(Sender: TObject; const OldLocation,
      NewLocation: TLocationCoord2D);
    procedure msAccelerometrSensorChoosing(Sender: TObject;
      const Sensors: TSensorArray; var ChoseSensorIndex: Integer);
    procedure tMotionTimer(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { Private declarations }
    FGeocoder : TGeocoder;
    procedure OnGeocodeReverseEvent(const Address: TCivicAddress);
    procedure OnGeocodeEvent(const Coords: TArray<TLocationCoord2D>);
  public
    { Public declarations }
  end;

var
  Form5: TForm5;

implementation

{$R *.dfm}

procedure TForm5.bSwitchClick(Sender: TObject);
begin
  if bSwitch.Caption = 'Turn On' then
  begin
    bSwitch.Caption := 'Turn off';
    lsLocation.Active := True;
    msAccelerometr.Active := True;
    osCompass.Active := True;
    osInclinometer.Active := True;
  end
  else
  begin
    bSwitch.Caption := 'Turn On';
    lsLocation.Active := False;
    msAccelerometr.Active := False;
    osCompass.Active := False;
    osInclinometer.Active := False;
  end;
end;

procedure TForm5.FormCreate(Sender: TObject);
begin
  FGeocoder := TGeocoder.Current.Create;
  FGeocoder.OnGeocodeReverse := OnGeocodeReverseEvent;
  FGeocoder.OnGeocode := OnGeocodeEvent;
end;

procedure TForm5.lsLocationLocationChanged(Sender: TObject; const OldLocation,
  NewLocation: TLocationCoord2D);
begin
  lCoordinates.Caption := Format('Latitude : %2.7f; Longitude : %2.7f', [NewLocation.Latitude, NewLocation.Longitude]);
  if FGeocoder <> nil then
  begin
    lAddress.Caption := 'Start geocoding';
    if FGeocoder.Supported then
      FGeocoder.GeocodeReverse(NewLocation)
    else
      lAddress.Caption := 'Geocoding not supporting';
  end
  else
    lAddress.Caption := 'Geocoder not found';
end;

procedure TForm5.msAccelerometrSensorChoosing(Sender: TObject;
  const Sensors: TSensorArray; var ChoseSensorIndex: Integer);
var
  I : integer;
begin
  for I := 0 to Length(Sensors) - 1 do
    if (Sensors[I] as TCustomMotionSensor).SensorType = TMotionSensorType.Accelerometer3D then
    begin
      ChoseSensorIndex := I;
      Break;
    end;
end;

procedure TForm5.OnGeocodeEvent(const Coords: TArray<TLocationCoord2D>);
begin
  if Length(Coords) > 0 then
    lGeoCoordinates.Caption := Format('After geocoding address. Lat : %2.7f, Long : %2.7f',[Coords[0].Latitude, Coords[0].Longitude]);
end;

procedure TForm5.OnGeocodeReverseEvent(const Address: TCivicAddress);
begin
  if Address <> nil then
  begin
    lAddress.Caption := 'After geocoding coords:' + Address.ToString;
    if FGeocoder <> nil then
      FGeocoder.Geocode(Address);
  end
  else
    lAddress.Caption := 'not found';
end;

procedure TForm5.osCompassSensorChoosing(Sender: TObject;
  const Sensors: TSensorArray; var ChoseSensorIndex: Integer);
var
  I : integer;
begin
  for I := 0 to Length(Sensors) - 1 do
    if (Sensors[I] as TCustomOrientationSensor).SensorType = TOrientationSensorType.Compass3D then
    begin
      ChoseSensorIndex := I;
      Break;
    end;
end;

procedure TForm5.osInclinometerSensorChoosing(Sender: TObject;
  const Sensors: TSensorArray; var ChoseSensorIndex: Integer);
var
  I : integer;
begin
  for I := 0 to Length(Sensors) - 1 do
    if (Sensors[I] as TCustomOrientationSensor).SensorType = TOrientationSensorType.Inclinometer3D then
    begin
      ChoseSensorIndex := I;
      Break;
    end;
end;

procedure TForm5.tMotionTimer(Sender: TObject);
begin
  if msAccelerometr.Sensor <> nil then
    lAccel.Caption := Format('AccelX : %-1.5f'#13#10'AccelY : %-1.5f'#13#10'AccelZ : %-1.5f',
      [msAccelerometr.Sensor.AccelerationX, msAccelerometr.Sensor.AccelerationY, msAccelerometr.Sensor.AccelerationZ] )
  else
    lAccel.Caption := '';
end;

procedure TForm5.tOrientationTimer(Sender: TObject);
begin
  if osCompass.Sensor <> nil then
    lCompass.Caption := Format('Heading : %3.1f',[osCompass.Sensor.CompMagHeading])
  else
    lCompass.Caption := '';
  if osInclinometer.Sensor <> nil then
    lInclinometr.Caption := Format('TiltX : %-3.5f '#13#10'TiltY : %-3.5f '#13#10'TiltZ : %-3.5f ',
      [osInclinometer.Sensor.TiltX, osInclinometer.Sensor.TiltY, osInclinometer.Sensor.TiltZ])
  else
    lInclinometr.Caption := '';
end;

end.

File: CPP/VCL/Sensors/uMain.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "uMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"
TfrmMain *frmMain;
//---------------------------------------------------------------------------
__fastcall TfrmMain::TfrmMain(TComponent* Owner)
	: TForm(Owner)
{
}
//---------------------------------------------------------------------------
#ifndef _WIN64
void __fastcall TfrmMain::OnGeocodeEvent(const _dt_System_Sensors_1 Coords)
#else /* _WIN64 */
void __fastcall TfrmMain::OnGeocodeEvent(const System::TArray__1<TLocationCoord2D> Coords)
#endif
{
	if (Coords.Length > 0) {
		lGeoCoordinates->Caption =
			lGeoCoordinates->Caption.sprintf(L"After geocoding address. Lat : %2.7f, Long : %2.7f",
			Coords[0].Latitude, Coords[0].Longitude);
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::OnGeocodeReverseEvent(TCivicAddress* const Address)
{
	if (Address != NULL) {
		lAddress->Caption = "After geocoding coords:" + Address->ToString();
		if (FGeocoder != NULL) {
            FGeocoder->Geocode(Address);
		}
	} else {
        lAddress->Caption = "not found";
    }
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::FormCreate(TObject *Sender)
{
	FGeocoder = (TGeocoder*)new TGeocoderClass(TGeocoder::Current);
	FGeocoder->OnGeocode = &OnGeocodeEvent;
	FGeocoder->OnGeocodeReverse = &OnGeocodeReverseEvent;
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::tOrientationTimer(TObject *Sender)
{
	if (osCompass->Sensor != NULL) {
		lCompass->Caption = lCompass->Caption.sprintf(L"Heading : %3.1f",
			osCompass->Sensor->CompMagHeading);
	} else {
		lCompass->Caption = "";
	}
	if (osInclinometer->Sensor != NULL) {
		lInclinometr->Caption = lInclinometr->Caption.sprintf(
			L"TiltX : %-3.5f \nTiltY : %-3.5f \nTiltZ : %-3.5f ",
			osInclinometer->Sensor->TiltX, osInclinometer->Sensor->TiltY,
			osInclinometer->Sensor->TiltZ);
	} else {
		lInclinometr->Caption = "";
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::tMotionTimer(TObject *Sender)
{
	if (msAccelerometr->Sensor != NULL) {
		lAccel->Caption = lAccel->Caption.sprintf(
			L"AccelX : %-1.5f\nAccelY : %-1.5f\nAccelZ : %-1.5f",
			msAccelerometr->Sensor->AccelerationX, msAccelerometr->Sensor->AccelerationY,
			msAccelerometr->Sensor->AccelerationZ);
	} else {
		lAccel->Caption = "";
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::osInclinometerSensorChoosing(TObject *Sender, const TSensorArray Sensors,
          int &ChoseSensorIndex)
{
	for (int i = 0; i < Sensors.Length; i++) {
		if (static_cast<TCustomOrientationSensor*>(Sensors[i])->SensorType ==
			TOrientationSensorType::Inclinometer3D) {
			ChoseSensorIndex = i;
			break;
		}
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::osCompassSensorChoosing(TObject *Sender, const TSensorArray Sensors,
          int &ChoseSensorIndex)
{
	for(int i = 0; i < Sensors.Length; i++)
	{
		if (static_cast<TCustomOrientationSensor*>(Sensors[i])->SensorType ==
			TOrientationSensorType::Compass3D) {
			ChoseSensorIndex = i;
			break;
		}
    }
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::msAccelerometrSensorChoosing(TObject *Sender, const TSensorArray Sensors,
          int &ChoseSensorIndex)
{
	for(int i = 0; i < Sensors.Length; i++) {
		if (static_cast<TCustomMotionSensor*>(Sensors[i])->SensorType ==
			TMotionSensorType::Accelerometer3D) {
			ChoseSensorIndex = i;
			break;
		}
    }
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::bSwitchClick(TObject *Sender)
{
	if (bSwitch->Caption == "Turn On") {
		bSwitch->Caption = "Turn off";
		lsLocation->Active = true;
		msAccelerometr->Active = true;
		osCompass->Active = true;
		osInclinometer->Active = true;
	} else {
		bSwitch->Caption = "Turn On";
		lsLocation->Active = false;
		msAccelerometr->Active = false;
		osCompass->Active = false;
		osInclinometer->Active = false;
	}
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::lsLocationLocationChanged(TObject *Sender, const TLocationCoord2D &OldLocation,
          const TLocationCoord2D &NewLocation)
{
	lCoordinates->Caption = lCoordinates->Caption.sprintf(
		L"Latitude : %2.7f; Longitude : %2.7f",
		NewLocation.Latitude, NewLocation.Longitude);
	if (FGeocoder != NULL) {
		lAddress->Caption = "Start geocoding";
		if (FGeocoder->Supported()) {
			FGeocoder->GeocodeReverse(NewLocation);
		} else {
			lAddress->Caption = "Geocoding not supporting";
		}
	} else {
        lAddress->Caption = "Geocoder not found";
    }

}
//---------------------------------------------------------------------------
